httpress.c

```diff
@@ -552,6 +548,70 @@ static double get_percentile(int p) // p - %
 	return idx * P_UPPER_BOUND / (double) P_RANGE;
 }
 
+static void clear_percentile(atomic_t* p)
+{
+	int i = 0;
+	for (i = 0; i <= P_RANGE; i++)
+		atomic_set(&p[i], 0);
+}
+
+static void reg_summary_percentile_val(double val)
+{
+	reg_percentile_val(val, percentile);
+}
+
+static double get_summary_percentile(int x)
+{
+	if (!(x > 0 && x < 100))
+		nxweb_die("Percentile value should be between 1 and 99");
+
+	return get_percentile(x, percentile);
+}
+
+/* percentile calculation stuff for intermediate reports */
+atomic_t tmp_percentile[P_RANGE+1];
+atomic_t lk_tmp_percentile;
+
+static void lock_tmp_percentile()
+{
+	__sync_or_and_fetch(&lk_tmp_percentile.val, 1);
+}
+
+static void unlock_tmp_percentile()
+{
+	__sync_and_and_fetch(&lk_tmp_percentile.val, 0);
+}
+
+static atomic_val_t is_locked_tmp_percentile()
+{
+	return atomic_get(&lk_tmp_percentile);
+}
+
+static void reg_tmp_percentile_val(double val)
+{
```

Comment:
With suche implementation of lock_tmp_percentile()/unlock_tmp_percentile() you still have a race condition here.

E.g.: is_locked_tmp_percentile() returns 0, and then before reg_percentile_val() is called another threads locks it.
If you want to do it atomically, you need to use CAS instruction (__sync_val_compare_and_swap() gcc intrinsic).